confuse/cache.py

```diff
@@ -0,0 +1,121 @@
+from typing import Dict, List
+
+from . import templates
+from .core import ROOT_NAME, Configuration, ConfigView, RootView, Subview
+
+
+class CachedHandle(object):
+    """Handle for a cached value computed by applying a template on the view.
+    """
+    # some sentinel objects
+    _INVALID = object()
+    _MISSING = object()
+
+    def __init__(self, view: ConfigView, template=templates.REQUIRED) -> None:
+        self.value = self._INVALID
+        self.view = view
+        self.template = template
+
+    def get(self):
+        """Retreive the cached value from the handle.
+
+        Will re-compute the value using `view.get(template)` if it has been
+        invalidated.
+
+        May raise a `NotFoundError` if the underlying view has been
+        invalidated.
+        """
+        if self.value is self._MISSING:
+            # will raise a NotFoundError if no default value was provided
+            self.value = templates.as_template(self.template).get_default_value()
+        if self.value is self._INVALID:
+            self.value = self.view.get(self.template)
+        return self.value
+
+    def _invalidate(self):
+        """Invalidate the cached value, will be repopulated on next `get()`.
+        """
+        self.value = self._INVALID
+
+    def _set_view_missing(self):
+        """Invalidate the handle, will raise `NotFoundError` on `get()`.
+        """
+        self.value = self._MISSING
+
+
+class CachedViewMixin:
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # keep track of all the handles from this view
+        self.handles: List[CachedHandle] = []
+        # need to cache the subviews to be able to access their handles
+        self.subviews: Dict[str, CachedConfigView] = {}
```

Comment:
I don't know if this is a good idea, but just to try it out here: would it simplify things at all to keep the list of handles _only on the `RootView`_? That is, any time any data changes anywhere in the entire configuration, we'd invalidate everything. I know this is possibly a bit more inefficient, but it would centralize the tracking of handles and avoid the need for all views to keep track of their subviews for invalidation purposes.

The reason I'm slightly nervous (perhaps unfoundedly) about the latter thing is that we have never before required that subviews be _unique_… that is, doing `config['foo']` twice in "normal" Confuse could give you back two different objects (equivalent objects, but different ones). These objects were short-lived and disposable; they generally did not stick around. Now we're saying that they must stick around, and in fact, that doing `config['foo']` twice _always_ gives you back exactly the same memoized view object. I'm just a tad worried that this means that, if there were some _other_ way of constructing a view object onto the same "place" in the configuration hierarchy (such as via iteration?), it would not enjoy the same invalidation benefits. So this seems to place a few more restrictions on the way that views can be created, stored, and used.

Reply:
I should mention that I could perhaps try prototyping this idea if it seems plausible to you!